### DISCUSSION TITLE
Read dynamic system environment variables from Kong lua plugin

### DISCUSSION THREAD

**Discussion by dsnr-coder** [Kong/kong#9104] [Help] [answered]
Hi All,

We have deployed kong in kubernetes cluster, we are trying to read kubernetes ip and port which was set to kong pod as an environment variable.

Is their any way we can read the environment variables from kong lua plugin...
We tried reading environment variables like below, but getting nil as response. os.getenv(\"KONG_NGINX_WORKER_PROCESSES\")

Thanks in advance for your help..

**Comment by flrgh** [comment, marked as answer]
Hi there :wave: 

NGINX/OpenResty starts its worker processes with an empty environment, so env vars set for the master process are not available within lua (with the exception of the `init` phase) [unless explicitly configured](https://github.com/openresty/openresty/issues/171) via NGINX's [env directive](https://nginx.org/en/docs/ngx_core_module.html#env).

Your plugin's `handler.lua` file is loaded during the `init` phase, so if there are any non-Kong environment vars you need to access, those can typically be accessed then:

<details>
<summary>handler.lua</summary>

```lua
local my_var = os.getenv("MY_ENV_VAR")

local my_plugin_handler = {
  PRIORITY = 100,
  VERSION = "0.1",
}

function my_plugin_handler:access()
  kong.log.info("my_var: ", my_var)
end

return my_plugin_handler
```
</details>


I wouldn't recommend this in your case though, because it's not a reliable way to access Kong configuration values (i.e. what if a value is set by `kong.conf` file and not by env var?). Kong configuration values are accessible on the global `kong.configuration` table:

```lua
local procs = kong.configuration.nginx_main_worker_processes
```

You can also use the [ngx.worker.count() function](https://github.com/openresty/lua-nginx-module#ngxworkercount) to retrieve the number of active worker processes. This is probably more helpful than reading from configuration, because `nginx_main_worker_processes` might be set to `auto`, in which case `ngx.worker.count()` is the only reliable way of checking the process count anyways:

```lua
local procs = ngx.worker.count()
```

**Reply by dsnr-coder** [reply]
Thanks a lot flrgh, the above solution worked for me, thanks for the detailed explanation, it worked for us

**Comment by MarkYQJ** [comment]
As @flrgh says, you have to passing it through NGINX's env directive.

Suppose the customized environment variable you defined in kong Deployment is
> `- name: MY_CAT_NAME`
> ` value: kitty`

You have to insert below code into nginx.conf

`env MY_CAT_NAME;`

Now, in lua plugin, you could read by `os.getenv("MY_CAT_NAME")`
It's more like telling NGINX `expose the environment variable MY_CAT_NAME`
The only thing left is to replace /usr/local/kong/nginx.conf in kong container _proxy_ with the modified nginx.conf.
